### ISSUE TITLE
How do I easily calculate the determinant of the Jacobian when it's a pytree of pytrees?

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#7183]
I would like to test that the determinant of the Fisher information of an exponential family is positive, but it's not easy to do that since the Jacobian returns a pytree of pytrees.  As it is, `jacfwd` and `jacrev` already calculate the Jacobian as an array before "unravelling" it into the pytree of pytrees. 

I'm not sure what the best design is, but these seem reasonable?

- I could try to flatten that pytree-of-pytrees into an array, which would mean stacking arrays horizontally and vertically using the treedefs?
- With an appropriate flag, the determinant could be calculated by `jacrev` or `jacfwd` and returned?
- With an appropriate flag, the Jacobian could be returned as an array, and I could call `jnp.linalg.det` on that?
- Maybe there's a way to calculate the determinant of a pytree-of-pytrees?

**Comment by jakevdp** [comment]
Thanks for the question - I'm having trouble understanding the context in which you're seeing jacobians return a pytree of pytrees. Can you show a concise example?

**Comment by NeilGirdhar** [comment]
@jakevdp I don't know how concise this is, but my application is simply to find the Fisher information [here](https://github.com/NeilGirdhar/efax/blob/master/efax/_src/natural_parametrization.py#L125), which is just
```python
jacfwd(grad(type(self).log_normalizer))
```
where `log_normalizer` is a scalar method of a distribution like [this one](https://github.com/NeilGirdhar/efax/blob/master/efax/_src/distributions/multivariate_normal/arbitrary.py#L24), essentially something like:
```python
@dataclass
class Normal:
    mean: RealArray
    variance: RealArray

    def log_normalizer(self) -> RealScalar:
        ...
```
The problem is that some distribution have both real and complex parameters.

**Comment by jakevdp** [comment]
Ah, I see. The issue is that you're computing the jacobian with respect to your dataclass type (`self` here). If you want to compute the jacobian with respect to an array, then you should pass `jacfwd` a function which takes that array as an argument, rather than a function that takes your dataclass as an argument. Does that make sense?

**Comment by NeilGirdhar** [comment]
@jakevdp Thank you, yes.  However, I don't just want the Fisher information of the individual parameters (which are the arrays).  I want the Fisher information of all of the parameters.  I think that's a reasonable thing to calculate.

**Comment by NeilGirdhar** [comment]
Although I think you just gave me an idea of how to work around this problem.  I already have functions to flatten and unflatten the structure.  I could define a `log_normalizer_flat` that works on the flattened version and calculate the Hessian of that.  Then, I would get the Fisher information as a matrix as desired.  So I guess that solves my problem !

**Comment by jakevdp** [comment]
I guess I'm still not understanding what's happening. For example, here is a simple version of compuing the forward jacobian of a pytree, and I get a pytree of results:
```python
import jax
import jax.numpy as jnp
from typing import NamedTuple

class Normal(NamedTuple):
  x: jnp.ndarray
  y: jnp.ndarray

  def func(self):
    return jnp.dot(self.x, self.y)

  gradfunc = jax.jacfwd(func)

val = Normal(jnp.arange(4.0), jnp.ones(4))
val.gradfunc()
# Normal(x=DeviceArray([1., 1., 1., 1.], dtype=float64), y=DeviceArray([0., 1., 2., 3.], dtype=float64))
```
Can you modify this example so that it results in the pytree-of-pytrees that you're seeing?

**Comment by NeilGirdhar** [comment]
@jakevdp I think what you've missed is that I'm calculating the Jacobian of the gradient?

**Comment by jakevdp** [comment]
Ah, I missed that, thanks.

What if you did something like this? Does the non-nested version give you the result you're interested in?
```python
import jax
import jax.numpy as jnp
from typing import NamedTuple

class Normal(NamedTuple):
  x: jnp.ndarray
  y: jnp.ndarray

  def func(self):
    return jnp.dot(self.x, self.y)

  gradfunc_nested = jax.jacfwd(jax.grad(func))
  
  def gradfunc_not_nested(self):
    def gradfunc(x, y):
      return self.__class__(x, y).func()
    return jax.jacobian(jax.grad(gradfunc))(self.x, self.y)

val = Normal(jnp.arange(4.0), jnp.ones(4))

print(val.gradfunc_nested())
# Normal(x=Normal(x=DeviceArray([[0., 0., 0., 0.],
#              [0., 0., 0., 0.],
#              [0., 0., 0., 0.],
#              [0., 0., 0., 0.]], dtype=float64), y=DeviceArray([[1., 0., 0., 0.],
#              [0., 1., 0., 0.],
#              [0., 0., 1., 0.],
#              [0., 0., 0., 1.]], dtype=float64)), y=Normal(x=DeviceArray([[1., 0., 0., 0.],
#              [0., 1., 0., 0.],
#              [0., 0., 1., 0.],
#              [0., 0., 0., 1.]], dtype=float64), y=DeviceArray([[0., 0., 0., 0.],
#              [0., 0., 0., 0.],
#              [0., 0., 0., 0.],
#              [0., 0., 0., 0.]], dtype=float64)))

print(val.gradfunc_not_nested())
# [[0. 0. 0. 0.]
#  [0. 0. 0. 0.]
#  [0. 0. 0. 0.]
#  [0. 0. 0. 0.]]
```

**Comment by NeilGirdhar** [comment]
I think your idea is along the lines of flattening and then taking the Hessian of the flattened structure.  However, in what you wrote `grad` is only taking the gradient of `x` since it defaults to taking the gradient of the first argument only.

**Comment by jakevdp** [comment]
You can pass `argnums=(0, 1)` if you want to take the gradient with respect to the first two arguments

**Comment by NeilGirdhar** [comment]
Yes, but then but then you'll have some kind of non-array structure again, right?

**Comment by jakevdp** [comment]
Right - computing the jacobian with respect to two parameters will give you a tuple of the two jacobians.

**Comment by NeilGirdhar** [comment]
It's just a shame because I know that the array I want is being calculated in `jacfwd` and `jacrev`.  But I should be able to get this working by flattening first and we can close this 😄   Thanks for your help.

**Comment by jakevdp** [comment]
Can you write down mathematically what you're hoping to compute? I'm not certain I've entirely grasped that yet.

**Comment by NeilGirdhar** [comment]
@jakevdp It's the Fisher information of an exponential family, which is equal to the Hessian of the log-normalizer.  If you look [here](https://github.com/NeilGirdhar/efax/blob/master/expfam.pdf), it's equations 53–56.

**Comment by jakevdp** [comment]
Ah, I see. Am I right in interpreting that $\eta$ in that equation is the vector of relevant parameters, which in your case would contain both the mean and the variance?

**Comment by NeilGirdhar** [comment]
@jakevdp Yes, you got it!

**Comment by jakevdp** [comment]
Ah, great!

In that case I'd recommend something along these lines:
```python
import jax
import jax.numpy as jnp
from typing import NamedTuple

class Normal(NamedTuple):
  x: jnp.ndarray
  y: jnp.ndarray

  def func(self):
    return jnp.dot(self.x ** 2 + self.y, self.y)

  def gradfunc(self):
    def wrapped(eta):
      x, y = eta[:len(eta) // 2], eta[len(eta) // 2:]
      return self.__class__(x, y).func()
    eta = jnp.concatenate([self.x, self.y])
    return jax.hessian(wrapped)(eta)

val = Normal(jnp.arange(4.0), jnp.ones(4))

val.gradfunc()
# DeviceArray([[2., 0., 0., 0., 0., 0., 0., 0.],
#              [0., 2., 0., 0., 0., 2., 0., 0.],
#              [0., 0., 2., 0., 0., 0., 4., 0.],
#              [0., 0., 0., 2., 0., 0., 0., 6.],
#              [0., 0., 0., 0., 2., 0., 0., 0.],
#              [0., 2., 0., 0., 0., 2., 0., 0.],
#              [0., 0., 4., 0., 0., 0., 2., 0.],
#              [0., 0., 0., 6., 0., 0., 0., 2.]], dtype=float64)
```
Since you're interested in the hessian with respect to the full parameter vector, you can compute your gradients from a function that accepts that vector.

**Comment by NeilGirdhar** [comment]
Yup, that makes sense!  Let's close this then.  Thank you for your help.
